File: backend/app/schemas/workflow.py
"""Schemas Pydantic para workflows de firma colaborativa.

Referencia: docs/API_SPEC.md §4
"""
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ===== Request =====
class SignerInput(BaseModel):
    cert_fingerprint: str
    name: Optional[str] = None
    email: Optional[str] = None


class WorkflowCreateRequest(BaseModel):
    signers: List[SignerInput] = Field(..., min_length=1)
    type: str = "parallel"  # parallel | sequential
    description: Optional[str] = None


# ===== Responses =====
class AssignmentCreated(BaseModel):
    id: int
    signer_cert_fingerprint: str
    status: str
    sequence_number: Optional[int] = None
    created_at: datetime


class WorkflowCreateResponse(BaseModel):
    workflow_id: int
    document_id: str
    type: str
    required_signers: int
    completed_signers: int
    status: str
    assignments: List[AssignmentCreated]
    created_at: datetime


class AssignmentSigner(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    first_name: Optional[str] = None
    last_name: Optional[str] = None


class AssignmentStatus(BaseModel):
    id: int
    signer: AssignmentSigner
    status: str
    sequence_number: Optional[int] = None
    signed_at: Optional[datetime] = None


class WorkflowStatusResponse(BaseModel):
    workflow_id: int
    document_id: str
    type: str
    status: str
    required_signers: int
    completed_signers: int
    assignments: List[AssignmentStatus]
    created_at: datetime
    completed_at: Optional[datetime] = None


# ===== Pendientes de firma (GET /auth/my-pending-signatures) =====
class PendingWorkflowInfo(BaseModel):
    workflow_id: int
    type: str
    completed_signers: int
    required_signers: int


class PendingCreatedBy(BaseModel):
    first_name: Optional[str] = None
    last_name: Optional[str] = None


class PendingSignatureItem(BaseModel):
    document_id: str
    title: str
    created_by: PendingCreatedBy
    created_at: datetime
    workflow: PendingWorkflowInfo


class PendingSignaturesResponse(BaseModel):
    total: int
    pending_signatures: List[PendingSignatureItem]

File: frontend/src/types/workflow.ts
export type WorkflowStatus = "pending" | "in_progress" | "completed" | "cancelled";
export type WorkflowSequenceType = "parallel" | "sequential";
export type AssignmentStatus = "pending" | "signed" | "rejected";

export interface SignerInput {
  cert_fingerprint: string;
  name?: string;
  email?: string;
}

export interface WorkflowCreateRequest {
  signers: SignerInput[];
  type: WorkflowSequenceType;
  description?: string;
}

export interface WorkflowAssignment {
  id: number;
  signer: { first_name?: string; last_name?: string };
  status: AssignmentStatus;
  sequence_number?: number;
  signed_at?: string;
}

export interface WorkflowStatusResponse {
  workflow_id: number;
  document_id: string;
  type: WorkflowSequenceType;
  status: WorkflowStatus;
  required_signers: number;
  completed_signers: number;
  assignments: WorkflowAssignment[];
  created_at: string;
  completed_at?: string;
}

export interface PendingSignatureItem {
  document_id: string;
  title: string;
  created_by: { first_name?: string; last_name?: string };
  created_at: string;
  workflow: {
    workflow_id: number;
    type: WorkflowSequenceType;
    completed_signers: number;
    required_signers: number;
  };
}

export interface PendingSignaturesResponse {
  total: number;
  pending_signatures: PendingSignatureItem[];
}
